execution/executor/src/lib.rs

```diff
@@ -336,50 +352,62 @@ where
-                .iter()
-                .any(|event| *event.key() == new_epoch_event_key)
-            {
-                let validator_set = account_to_state
-                    .get(&on_chain_config::config_address())
-                    .map(|state| {
-                        state
-                            .get_validator_set()?
-                            .ok_or_else(|| format_err!("ValidatorSet does not exist"))
-                    })
-                    .ok_or_else(|| format_err!("ValidatorSet account does not exist"))??;
-                let configuration = account_to_state
-                    .get(&on_chain_config::config_address())
-                    .map(|state| {
-                        state
-                            .get_configuration_resource()?
-                            .ok_or_else(|| format_err!("Configuration does not exist"))
-                    })
-                    .ok_or_else(|| format_err!("Association account does not exist"))??;
-                Some(EpochState {
-                    epoch: configuration.epoch(),
-                    verifier: (&validator_set).into(),
+        // check for change in validator set
+        let next_epoch_state = if new_epoch_marker.is_some() {
+            // Pad the rest of transactions
+            txn_data.resize(
+                transactions.len(),
+                TransactionData::new(
+                    HashMap::new(),
+                    vec![],
+                    TransactionStatus::Retry,
+                    current_state_tree.root_hash(),
+                    Arc::new(InMemoryAccumulator::<EventAccumulatorHasher>::default()),
+                    0,
+                    None,
+                ),
+            );
+
+            let validator_set = account_to_state
+                .get(&on_chain_config::config_address())
+                .map(|state| {
+                    state
+                        .get_validator_set()?
+                        .ok_or_else(|| format_err!("ValidatorSet does not exist"))
                 })
-            } else {
-                None
-            }
-        }
+                .ok_or_else(|| format_err!("ValidatorSet account does not exist"))??;
+            let configuration = account_to_state
+                .get(&on_chain_config::config_address())
+                .map(|state| {
+                    state
+                        .get_configuration_resource()?
+                        .ok_or_else(|| format_err!("Configuration does not exist"))
+                })
+                .ok_or_else(|| format_err!("Association account does not exist"))??;
+            Some(EpochState {
+                epoch: configuration.epoch(),
+                verifier: (&validator_set).into(),
+            })
+        } else {
+            None
+        };
 
         let current_transaction_accumulator =
             parent_trees.txn_accumulator().append(&txn_info_hashes);
+
+        current_state_tree.prune();
```

Comment:
The trees used to be pruned in the commit stage. Although data structure changed due to this diff, pruning them in the execute stage definitely looks wrong.

Reply:
I agree, I think what was happening before is that execution of each block adds it's speculative updates to the tree, and when blocks get committed, those states are now in the storage so we prune the in-memory tree by dropping those nodes (if we need them again the subtree-proofs will contain what's the latest committed). Pruning here is wrong imho because the state isn't yet committed.